lib/lev/status.rb

```diff
@@ -0,0 +1,131 @@
+require 'json'
+
+module Lev
+  class Status
+    STATUS_QUEUED = 'queued'
+    STATUS_WORKING = 'working'
+    STATUS_COMPLETED = 'completed'
+    STATUS_FAILED = 'failed'
+    STATUS_KILLED = 'killed'
+
+    STATUSES = [
+      STATUS_QUEUED,
+      STATUS_WORKING,
+      STATUS_COMPLETED,
+      STATUS_FAILED,
+      STATUS_KILLED
+    ].freeze
+
+    attr_reader :uuid
+
+    def initialize(uuid = nil)
+      @uuid = uuid || SecureRandom.uuid
+      save
+    end
+
+    def self.find(uuid)
+      if status = store.fetch(status_key(uuid))
+        JSON.parse(status)
+      else
+        nil
+      end
+    end
+
+    def set_progress(at, out_of = nil)
+      prevent_faulty_arguments(at, out_of)
+
+      if set_status_progress(at, out_of) == 1.0
+        set(status: STATUS_COMPLETED)
+      end
+    end
+
+    STATUSES.each do |status|
+      define_method("#{status}!") do
+        set(status: status)
+      end
+    end
+
+    def save(hash = {})
+      if has_reserved_keys?(hash)
+        raise IllegalArgument,
+              "Caller cannot specify any reserved keys (#{RESERVED_KEYS})"
+      else
+        set(hash)
+      end
+    end
+
+    def add_error(error, options = { })
+      options = { is_fatal: false }.merge(options)
+      push('errors', { is_fatal: options[:is_fatal],
+                       code: error.code,
+                       message: error.message })
+    end
+
+    def get(key)
+      self.class.find(uuid)[key]
+    end
+
+    protected
+    RESERVED_KEYS = [:progress, :uuid, :status, :errors]
+
+    def self.store
+      # Nice to get the store from lev config each time so it isn't serialized
+      # when activejobs are sent off to places like redis
+      Lev.configuration.status_store
+    end
+
+    def set(incoming_hash)
+      if existing_settings = self.class.find(uuid)
+        incoming_hash = existing_settings.merge(incoming_hash)
+      end
+
+      self.class.store.write(status_key, incoming_hash.to_json)
+    end
+
+    def status_key
+      "#{Lev.configuration.status_store_namespace}:#{uuid}"
+    end
+
+    def self.status_key(uuid)
+      "#{Lev.configuration.status_store_namespace}:#{uuid}"
+    end
+
+    def has_reserved_keys?(hash)
+      (hash.keys.collect(&:to_sym) & RESERVED_KEYS).any?
+    end
+
+    def push(key, new_item)
+      new_value = (get(key) || []).push(new_item)
+      set(key => new_value)
+    end
+
+    STATUSES.each do |status|
+      define_method("#{status}?") do
+        get('status') == status
+      end
+    end
+
+    def set_status_progress(at, out_of)
+      if out_of.nil? && (at < 0 || at > 1)
```

Comment:
This condition looks like it belongs in `prevent_faulty_arguments`?  Can maybe the `prevent_faulty_arguments` checks be moved to `set_status_progress`?  Or if not, change the name of `prevent_faulty_arguments` to be progress-specific?